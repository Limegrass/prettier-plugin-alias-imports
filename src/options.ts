import { ParserOptions, SupportOption } from "prettier";

export const SupportedParsers = {
    TypeScript: "typescript",
    Babel: "babel",
    Flow: "flow",
} as const;

/** Type for the values of `SupportedParsers`. */
export type SupportedParserName =
    typeof SupportedParsers[keyof typeof SupportedParsers];

export interface AliasImportOptions extends ParserOptions {
    aliasConfigPath?: string;
    parser: SupportedParserName;
}

/**
 * Force `options` to only include keys from the AliasImportOptions interface
 */
type AliasImportSupportOptions = Omit<
    Record<keyof AliasImportOptions, SupportOption>,
    keyof ParserOptions
>;

/**
 * Configuration options specific to this plugin
 * which prettier presents to the user if installed.
 */
export const options: AliasImportSupportOptions = {
    // Would be nice to be able to use some type
    // type ConfigFile = { path: string; type: SupportedParserPlugins }
    // but I don't think that is supported
    aliasConfigPath: {
        since: "0.1.0",
        type: "path", // string is supported by prettier but the types needs to be updated
        category: "Global",
        description: "Configuration file to read. Defaults to tsconfig.json",
    },
};
