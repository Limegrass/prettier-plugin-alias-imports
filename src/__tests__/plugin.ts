import aliasPlugin from "#src/plugin";
import { parsers as babelParsers } from "prettier/parser-babel";
import { parsers as flowParsers } from "prettier/parser-flow";
import { parsers as typescriptParsers } from "prettier/parser-typescript";
import { format } from "prettier";
import { SupportedParsers } from "#src/options";

const code = `
    import { write } from "fs";
    import { potato } from "./potato";
`;

const fakeCodeAddition = "// potato";
const fakePreprocess = (code: string) => code + fakeCodeAddition;

describe("prettier-plugin-alias-imports", () => {
    it.each([
        {
            parserName: SupportedParsers.Babel,
            fileExtension: "js",

            parser: babelParsers.babel,
        },
        {
            parserName: SupportedParsers.Flow,
            fileExtension: "js",
            parser: flowParsers.flow,
        },
        {
            parserName: SupportedParsers.TypeScript,
            fileExtension: "ts",
            parser: typescriptParsers.typescript,
        },
    ])(
        "formats with other $parserName plugins",
        ({ parserName, parser, fileExtension }) => {
            const formattedCode = format(code, {
                filepath: `./src/plugin.${fileExtension}`,
                parser: parserName,
                plugins: [
                    {
                        parsers: {
                            [parserName]: {
                                ...parser,
                                preprocess: fakePreprocess,
                            },
                        },
                    },
                    aliasPlugin, // should be last
                ],
            });
            expect(formattedCode).toMatch(fakeCodeAddition);
            expect(formattedCode).toMatch(`#src/potato`);
            expect(formattedCode).toMatchSnapshot(); // purely for inspection
        }
    );
});
