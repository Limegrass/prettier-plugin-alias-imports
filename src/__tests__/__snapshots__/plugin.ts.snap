// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prettier-plugin-alias-imports formats with other babel plugins 1`] = `
"import { write } from \\"fs\\";
import { potato } from \\"#src/potato\\";
// potato
"
`;

exports[`prettier-plugin-alias-imports formats with other flow plugins 1`] = `
"import { write } from \\"fs\\";
import { potato } from \\"#src/potato\\";
// potato
"
`;

exports[`prettier-plugin-alias-imports formats with other typescript plugins 1`] = `
"import { write } from \\"fs\\";
import { potato } from \\"#src/potato\\";
// potato
"
`;
